pyrealm/demography/canopy_functions.py

```diff
@@ -0,0 +1,411 @@
+    """
+    # Scaling factor to give expected A_c (crown area) at
+    # z_m (height of maximum crown radius)
+
+    return 1 / q_m * np.sqrt(crown_area / np.pi)
+
+
+def _validate_z_args(z: NDArray[np.float32], *args: NDArray[np.float32]) -> None:
+    """Shared validation routine for vertical height arguments.
+
+    The canopy functions that take a height argument ``z`` can take a range of input
+    shapes for ``z``, alongside a set of other row arrays representing cohort
+    properties. This function is used to check that the cohort properties are of equal
+    length and that the ``z`` value has one of the accepted shapes described in
+    :meth:`~pyrealm.demography.canopy_functions.calculate_relative_canopy_radius_at_z`.
+
+    Args:
+        z: The inputs to the ``z`` argument of a function.
+        args: Other arrays representing cohort properties.
+    """
+
+    if z.size == 1 or z.ndim == 1:
+        # All same height or stem specific heights - check z is either a scalar or also
+        # a row vector of the right length.
+        check_input_shapes(z, *args)
+        return
+    elif z.ndim == 2 and z.shape[1] == 1:
+        # Z is a column vector, just check stem properties.
+        check_input_shapes(*args)
+        return
+
+    raise ValueError("Invalid shape for the z value.")
+
+
+def _validate_q_z(
+    z: NDArray[np.float32],
+    q_z: NDArray[np.float32],
+    stem_property: NDArray[np.float32],
+) -> None:
+    """Shared validation routine for relative radius arguments.
+
+    The functions
+    :meth:`~pyrealm.demography.canopy_functions.calculate_stem_projected_canopy_area_at_z`
+    and
+    :meth:`~pyrealm.demography.canopy_functions.calculate_stem_projected_leaf_area_at_z`
+    both require the arguments ``z`` and ``q_z``, where ``z`` is an array of vertical
+    heights and ``q_z`` is the relative canopy radius at those heights for a set of
+    stems. This function checks that the inputs are congruent with each other, and with
+    the shape of a stem property argument,  given the set of expected forms of ``z``
+    described in
+    :meth:`~pyrealm.demography.canopy_functions.calculate_relative_canopy_radius_at_z`.
+
+
+    Args:
+        z: The input to the ``z`` argument.
+        q_z: The input to the ``q_z`` argument.
+        stem_property: An argument input representing a stem property.
+    """
+
+    if z.size == 1 or z.ndim == 1:
+        # All same height or stem specific heights - q_z must then also be row vector of
+        # the same length as the stem propertys and z must be scalar or a row_vector.
+        check_input_shapes(z, q_z, stem_property)
+        return
+    elif z.ndim == 2 and z.shape[1] == 1:
+        # z is a column array, so check q_z is a matrix
+        if q_z.shape != (z.size, stem_property.size):
+            raise ValueError("Invalid shape for q_z.")
+        return
+
+    raise ValueError("Invalid shape for the z value.")
+
+
+def calculate_relative_canopy_radius_at_z(
```

Comment:
If the inputs array can be reshaped to be broadcastable, then using numpy functions instead of python built-ins e.g. `np.power` instead of `**`can help with vectorization and improve performance.

Reply:
I need to think about (and document) what shapes are sensible here. I believe that `M**n` is faster if `n` is a scalar but `np.power(M, n)` is faster when `n` is an array? Could get either in using this function.

Reply:
I think np.power will be faster if you are applying the operation to an array even if the exponent is a scaler. And you can further performance gains if you can apply np.vectorize to functions. Though I feel like I need to back up this with some testing!